src/backend/src/lib.rs

```diff
@@ -126,6 +135,8 @@ pub struct State {
     /// Introduced to support a broader range of user-defined custom tokens, beyond just ERC20.
     /// Future updates may include migrating existing ERC20 tokens to this more flexible structure.
     custom_token: CustomTokenMap,
+    user_profile: UserProfileMap,
+    user_profile_updated: UserProfileUpdatedMap,
```

Comment:
Maybe I'm dumb but, I dont not understand why there are two Maps?

Reply:
The key of the first map is a tuple of updated and principal. That means I can't select only based on principal. But it allows me to select by updated, which will be necessary for get_users.

At the same time, we can't use only the principal as key, because due to the cycle consumption, we might not be able to iterate through the whole map to filter per updated at.

In the end the solution was two maps:
* One to get profiles filtered by updated.
* One to get the updated by principal, and then use the pair to get the user profile from the other map.

Does that make sense?

Reply:
> One to get profiles filtered by updated.

Thanks, I understand the logic now. However, I'm wondering for which features we need to get a list of users by timestamp?

Reply:
For the Rewards canister that Frederik is working on.

Reply:
I see. Now I understand the features but, I'm not sure about the proposed structure of the information I have to say.

Currently those are the maps defined in the PR:

```
type UserProfileMap = StableBTreeMap<(Timestamp, StoredPrincipal), Candid<StoredUserProfile>, VMem>;
type UserProfileUpdatedMap = StableBTreeMap<StoredPrincipal, Timestamp, VMem>;
```

This works fine but, I'm not sure it serves best the first goal of the feature - having user profiles.

For me, the main required features are:

- On sign-in, a profile is resolved for the user principal
- If no principal exists, a profile is created
- Maybe in the future, user can edit their profile in the settings page, for example setting a nickname

And additionally, for gameficication reasons, we also need

- A list of the user principals which completed the verification after a time x

For the main features, I got the feeling that it would probably be performant to have a map principal <> profile

```
type UserProfileMap = StableBTreeMap<StoredPrincipal, Candid<StoredUserProfile>, VMem>;
```

That way, on each login there could be a single access to a map and any additional requests are also limited to getting the information directly for the caller.

Regarding the additional functions, depends if we are fine duplicating or not the information but, I feel like it's more an "historical" data map than a profile related information.

```
type VerifiedUserProfileHistoryMap = StableBTreeMap<Timestamp, StoredPrincipal, VMem>;
// or
type VerifiedUserProfileHistoryMap = StableBTreeMap<Timestamp, Candid(<StoredRelevantInformation>), VMem>;
```

Not sure if I explain really well what I mean and if it makes sense?

Reply:
Those are good points, but I think it's better to do something more generic than storing the data needed by that endpoint now. Moreover, using only the Timestamp by key could be problematic if two share the same timestamp. Very unlikely, but I don't think it's worth the risk.

I believe that @bitdivine already has a use case to retrieve more user data. So, with my proposed implementation, a new endpoint could provide whatever is needed easily and without migrating.

I shared the architecture before implementing and we all discussed it there. Maybe you missed it: https://docs.google.com/document/d/1R-VFUzHsXLewSqMxEpEH3kDWOGdPUQcwGP4Ia5ucB48/edit#heading=h.b9uawa2tmu48

The current architecture is also very performant for the main feature. It just requires two accesses instead of one, but both are constant time.

Reply:
>  Moreover, using only the Timestamp by key could be problematic 

Agree.

> Maybe you missed it

Absolutely.

> It just requires two accesses instead of one, but both are constant time.

Personally, I think it's a cost we should spare because the main goal does not requires it or in other words because it deserves the main goal - i.e. I think we should design firstly the feature for the main feature - i.e. user needs to have profile that are retrieved on sign-in - and implement additional features around it.

That said, of course if Max and you already agree on this architecture, definitely fine for me!

Reply:
I don't think it's such a big cost, and it's more flexible than copying the data for the current use case.

Reply:
Like I said as you wish.

Reply:
By the way, user data is also in say the custom token map.  So it's not as if all the user profile data is in this structure.  Is it worth making the name reflect the actual content, e.g. `user_credentials`?

Reply:
My feeling is that this works, and while there are a lot of policy decisions that we might change, it isn't a big deal to migrate the data to another structure should we wish to at some point.

Reply:
I'd rather keep it as UserProfile because the point is to add more fields to it and not have it only for credentials.

That was recommended and agreed with @peterpeterparker